### PR TITLE
Lastcheque endpoints

### PR DESCRIPTION
**Pull request by metacertain**
The aim of this pr is to add the 
/chequebook/lastcheque/{peer}
&
/chequebook/lastcheque
endpoints on the debug api.

### CODE REVIEW

File: pkg/settlement/swap/swap.go [hunk 1]
```diff
@@ -212,3 +219,71 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
 	}
 	return nil
 }
+
+func (s *Service) LastChequePeer(peer swarm.Address) (*chequebook.SignedCheque, error) {
```
Comment (ralph-pichler) [inline pkg/settlement/swap/swap.go:223]: `LastSentCheque`

File: pkg/settlement/swap/swap.go [hunk 2]
```diff
@@ -212,3 +219,71 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
...
+	}
+
+	if !known {
+		return nil, ErrUnknownBeneficary
+	}
+
+	return s.chequebook.LastCheque(common)
+}
+
+func (s *Service) LastStoredChequePeer(peer swarm.Address) (*chequebook.SignedCheque, error) {
```
Comment (ralph-pichler) [inline pkg/settlement/swap/swap.go:238]: `LastReceivedCheque`

File: pkg/settlement/swap/swap.go [hunk 3]
```diff
@@ -212,3 +219,71 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
...
+	}
+
+	if !known {
+		return nil, ErrUnknownBeneficary
+	}
+
+	return s.chequeStore.LastCheque(common)
+}
+
+func (s *Service) LastCheques() (map[string]*chequebook.SignedCheque, error) {
```
Comment (ralph-pichler) [inline pkg/settlement/swap/swap.go:253]: `LastSentCheques`

File: pkg/settlement/swap/swap.go [hunk 4]
```diff
@@ -212,3 +219,71 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
...
+		if err == nil && known {
+			resultmap[addr.String()] = j
+		}
+	}
+
+	return resultmap, nil
+
+}
+
+func (s *Service) LastStoredCheques() (map[string]*chequebook.SignedCheque, error) {
```
Comment (ralph-pichler) [inline pkg/settlement/swap/swap.go:272]: `LastReceivedCheques`

**Review by ralph-pichler — APPROVED** [review]
There are a bunch of excessive new lines, but the logic seems sound and has been both unit and manually tested.
PR description is old, the endpoints are now `/chequebook/cheque/{peer}` and `/chequebook/cheque`.

File: pkg/debugapi/chequebook.go [hunk 5]
```diff
@@ -48,3 +70,117 @@ func (s *server) chequebookAddressHandler(w http.ResponseWriter, r *http.Request
...
+	}
+
+	var lastsent *chequebookLastChequePeerResponse
+	if err == nil {
+		lastsent = &chequebookLastChequePeerResponse{
+			Beneficiary: lastSent.Cheque.Beneficiary.String(),
+			Chequebook:  lastSent.Cheque.Chequebook.String(),
+			Payout:      lastSent.Cheque.CumulativePayout,
+		}
+	}
```
Comment (janos) [inline pkg/debugapi/chequebook.go:107]: Why not to put this before LastReceivedCheque call and avoid err2 and larger distance between handling these errors?

File: pkg/debugapi/chequebook.go [hunk 6]
```diff
@@ -48,3 +70,117 @@ func (s *server) chequebookAddressHandler(w http.ResponseWriter, r *http.Request
...
+}
+
+func (s *server) chequebookAllLastHandler(w http.ResponseWriter, r *http.Request) {
+
+	lastchequessent, err := s.Swap.LastSentCheques()
+
+	if err != nil {
+		jsonhttp.InternalServerError(w, errCantLastCheque)
+		return
+	}
```
Comment (janos) [inline pkg/debugapi/chequebook.go:132]: Log this error.

File: pkg/debugapi/chequebook.go [hunk 7]
```diff
@@ -48,3 +70,117 @@ func (s *server) chequebookAddressHandler(w http.ResponseWriter, r *http.Request
...
+
+	lastchequesreceived, err := s.Swap.LastReceivedCheques()
+
+	if err != nil {
+		jsonhttp.InternalServerError(w, errCantLastCheque)
+		return
+	}
```
Comment (janos) [inline pkg/debugapi/chequebook.go:139]: Log error.

File: pkg/settlement/swap/swap.go [hunk 8]
```diff
@@ -212,3 +223,73 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
...
+	return s.chequebook.LastCheque(common)
+}
+
+// LastReceivedCheque returns the last received cheque for the peer
+func (s *Service) LastReceivedCheque(peer swarm.Address) (*chequebook.SignedCheque, error) {
+
+	common, known, err := s.addressbook.Chequebook(peer)
+
+	if err != nil {
+		return nil, errors.New("error translating swarm address to chequebook address")
```
Comment (janos) [inline pkg/settlement/swap/swap.go:249]: Word error is redundant in the error message and the original error is not propagated.

File: pkg/settlement/swap/swap.go [hunk 9]
```diff
@@ -212,3 +223,73 @@ func (s *Service) Handshake(peer swarm.Address, beneficiary common.Address) erro
...
 	return nil
 }
+
+// LastSentCheque returns the last sent cheque for the peer
+func (s *Service) LastSentCheque(peer swarm.Address) (*chequebook.SignedCheque, error) {
+
+	common, known, err := s.addressbook.Beneficiary(peer)
+
+	if err != nil {
+		return nil, errors.New("error translating swarm address to chequebook address")
```
Comment (janos) [inline pkg/settlement/swap/swap.go:233]: Word error is redundant in the error message and the original error is not propagated.

File: pkg/debugapi/chequebook_test.go [hunk 10]
```diff
@@ -121,3 +125,274 @@ func TestChequebookAddress(t *testing.T) {
 	}
 
 }
+
+func TestChequebookLastCheques(t *testing.T) {
+
```
Comment (paxthemax) [inline pkg/debugapi/chequebook_test.go:130]: Remove newlines?

**Review by paxthemax — COMMENTED** [review]
One outstanding issue, everything else cosmetic. Good job in such a short time!

File: pkg/debugapi/chequebook.go (hunk 5 above)
Comment (metacertain) [reply pkg/debugapi/chequebook.go]: refactored this partly with the order proposed here

File: pkg/debugapi/chequebook.go (hunk 6 above)
Comment (metacertain) [reply pkg/debugapi/chequebook.go]: logged

File: pkg/debugapi/chequebook.go (hunk 7 above)
Comment (metacertain) [reply pkg/debugapi/chequebook.go]: logged

File: pkg/settlement/swap/swap.go (hunk 8 above)
Comment (metacertain) [reply pkg/settlement/swap/swap.go]: changed to return nil, err

File: pkg/settlement/swap/swap.go (hunk 9 above)
Comment (metacertain) [reply pkg/settlement/swap/swap.go]: changed to return nil, err

File: pkg/debugapi/chequebook.go [hunk 11]
```diff
@@ -48,3 +70,117 @@ func (s *server) chequebookAddressHandler(w http.ResponseWriter, r *http.Request
...
+
+	lastReceived, err2 := s.Swap.LastReceivedCheque(peer)
+	if err2 != nil && err2 != chequebook.ErrNoCheque {
+		s.Logger.Debugf("debug api: chequebook lastcheque peer: get peer %s last cheque: %v", peer.String(), err2)
+		s.Logger.Errorf("debug api: chequebook lastcheque peer: can't get peer %s last cheque", peer.String())
+		jsonhttp.InternalServerError(w, errCantLastChequePeer)
+		return
+	}
+
+	var lastsent *chequebookLastChequePeerResponse
```
[the remark this answers is not in the document]
Comment (metacertain) [reply pkg/debugapi/chequebook.go]: renamed to lastSentResponse

File: pkg/debugapi/chequebook.go [hunk 12]
```diff
@@ -48,3 +70,117 @@ func (s *server) chequebookAddressHandler(w http.ResponseWriter, r *http.Request
...
+	if err == nil {
+		lastsent = &chequebookLastChequePeerResponse{
+			Beneficiary: lastSent.Cheque.Beneficiary.String(),
+			Chequebook:  lastSent.Cheque.Chequebook.String(),
+			Payout:      lastSent.Cheque.CumulativePayout,
+		}
+	}
+
+	var lastreceived *chequebookLastChequePeerResponse
```
[the remark this answers is not in the document]
Comment (metacertain) [reply pkg/debugapi/chequebook.go]: renamed to lastReceivedResponse

**Comment by ralph-pichler** [comment]
@metacertain I added the openapi spec for the calls introduced here to the cashout PR building on top of this (because it introduces the EthereumAddress type there which would also be needed here).

**Review by zelig — APPROVED** [review]

**Pull request merged**